src/DGtal/topology/HalfEdgeDataStructure.h

```diff
@@ -0,0 +1,633 @@
+// Inclusions
+#include <iostream>
+#include <array>
+#include "DGtal/base/Common.h"
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+
+  /// Defines the invalid half-edge (i.e. exterior).
+  ///
+  /// @note It is defined external to the HalfEdgeDataStructure
+  /// because there was link error with gcc 4.8.4 for all variants
+  /// within HalfEdgeDataStructure (e.g. BOOST_STATIC_CONSTANT, c++
+  /// constexpr, c++ constexpr with const reference).
+  static std::size_t const HALF_EDGE_INVALID_INDEX = boost::integer_traits<std::size_t>::const_max;
+  
+  /////////////////////////////////////////////////////////////////////////////
+  // class HalfEdgeDataStructure
+  /**
+   * Description of template class 'HalfEdgeDataStructure' <p> \brief
+   * Aim: This class represents an half-edge data structure, which is
+   * a structure for representing the topology of a combinatorial
+   * 2-dimensional surface or an embedding of a planar graph in the
+   * plane. It does not store any geometry. As a minimal example,
+   * these lines of code build two triangles connected by the edge
+   * {1,2}.
+   *
+   * \code
+   * std::vector< HalfEdgeDataStructure::Triangle > triangles( 2 );
+   * triangles[0].v = { 0, 1, 2 };
+   * triangles[1].v = { 2, 1, 3 };
+   * HalfEdgeDataStructure mesh;
+   * mesh.build( triangles );
+   * std::cout << mesh << std::endl;
+   * \endcode
+   *
+   * @todo For now, the construction of the half-edge structure
+   * (method HalfEdgeDataStructure::build) is limited to triangles. It
+   * would be not difficult to adapt it to cellular surfaces.
+   * 
+   * @note Large parts of this class are taken from
+   * https://github.com/yig/halfedge, written by Yotam Gingold.
+   */
+  class HalfEdgeDataStructure
+  {
+  public:
+
+    /// The type for counting elements.
+    typedef std::size_t    Size;
+    /// The type used for numbering half-edges (an offset an the half-edges structure).
+    typedef std::size_t    Index;
+    /// The type used for numbering half-edges (alias)
+    typedef Index   HalfEdgeIndex;
+    /// The type for numbering vertices
+    typedef Index   VertexIndex;
+    /// The type for numbering edges
+    typedef Index   EdgeIndex;
+    /// The type for numbering faces
+    typedef Index   FaceIndex;
+
+    // Defines the invalid index.
+    // JOL: works with Apple LLVM version 8.1.0 (clang-802.0.42), but does not work with gcc 4.8.4
+    //
+    // static constexpr Index const& INVALID_INDEX = boost::integer_traits<Index>::const_max;
+    //
+    // JOL: does NOT work with Apple LLVM version 8.1.0 (clang-802.0.42)
+    // (undefined reference at link time).
+    // static constexpr Index const INVALID_INDEX = boost::integer_traits<Index>::const_max;
+    // static Index const INVALID_INDEX = boost::integer_traits<Index>::const_max;
+    // BOOST_STATIC_CONSTANT( Index, INVALID_INDEX = boost::integer_traits<Index>::const_max );
```

Comment:
strange because be have plenty of ```BOOST_STATIC_CONSTANT``` in other classes...

Reply:
I think I just understood what is the problem. It is that HalfEdgeDataStructure is not a template class. So BOOST_STATIC_CONSTANT seems not to work properly for static constants within normal classes. I wonder if it useful to dwell more into that issue.